File: preparation_2.pl
#Copyright (c) 2025. RIKEN All rights reserved.
#This is for academic and non-commercial research use only.
#The technology is currently under patent application.
#Commercial use is prohibited without a separate license agreement.
#E-mail: akihiro.ezoe@riken.jp

#!/usr/bin/perl
use strict;
use warnings;

sub calculate_stats {
    my @values = sort { $a <=> $b } @_;
    my $n = scalar @values;

    if( $n == 0 ){
	$n = 1;
    }
    
    my $sum = 0;
    $sum += $_ for @values;
    my $mean = $sum / $n;
    
    my $q1_idx = int($n / 4);
    my $q2_idx = int($n / 2);
    my $q3_idx = int(3 * $n / 4);
    
    my $q1 = $values[$q1_idx];
    my $q2 = $values[$q2_idx];
    my $q3 = $values[$q3_idx];
    
    return ($mean, $q1, $q2, $q3);
}

sub calculate_slope {
    my ($x1, $y1, $x2, $y2) = @_;
    return ($y2 - $y1) / ($x2 - $x1) if $x2 != $x1;
    return 0;
}


my $input_file = 'output_EZ/clustered_data.txt';
open my $fh, '<', $input_file or die "Cannot open file: $!";
my $header = <$fh>;
my @data;
while (my $line = <$fh>) {
    chomp $line;
    my @fields = split /\s+/, $line;
    push @data, \@fields;
}
close $fh;


my %groups;
foreach my $row (@data) {
    my ($x, $y, $cluster) = @$row;
    push @{$groups{$cluster}}, [$x, $y];
}


my %stats;
foreach my $cluster (keys %groups) {

    
    my @x_values = map { $_->[0] } @{$groups{$cluster}};
    my @y_values = map { $_->[1] } @{$groups{$cluster}};
    
    my ($x_mean, $x_q1, $x_q2, $x_q3) = calculate_stats(@x_values);
    my ($y_mean, $y_q1, $y_q2, $y_q3) = calculate_stats(@y_values);
    
    my @slopes;
    for my $i (0..$#{$groups{$cluster}}-1) {
        for my $j (($i+1)..$#{$groups{$cluster}}) {
            my $x1 = $groups{$cluster}[$i][0];
            my $y1 = $groups{$cluster}[$i][1];
            my $x2 = $groups{$cluster}[$j][0];
            my $y2 = $groups{$cluster}[$j][1];
            
            next if $x1 == $x2;
            
            my $slope = calculate_slope($x1, $y1, $x2, $y2);
            push @slopes, $slope;
        }
    }

    my ($slope_mean, $slope_q1, $slope_q2, $slope_q3) = calculate_stats(@slopes);
    my $slo = scalar @slopes;

    if ( $slo == 0 || $slo == 1 ){
	$slope_mean = $y_mean - $x_mean;
	$slope_q1 = $y_q1 - $x_q1;
	$slope_q2 = $y_q2 - $x_q2;
	$slope_q3 = $y_q3 - $x_q3;
    }

    my $m = @{$groups{$cluster}};
        
    $stats{$cluster} = {
        x => {
            mean => $x_mean,
            q1 => $x_q1,
            q2 => $x_q2,
            q3 => $x_q3,
	    n => $m
        },
        y => {
            mean => $y_mean,
            q1 => $y_q1,
            q2 => $y_q2,
            q3 => $y_q3,
	    n => $m
        },
        slope => {
            mean => $slope_mean,
            q1 => $slope_q1,
            q2 => $slope_q2,
            q3 => $slope_q3,
	    n => $m
        }
    };
    
    my $y_intercept = $y_mean - $slope_mean * $x_mean;
    my $calc_y_mean = $y_intercept + $slope_mean * $x_mean;
    my $calc_y_median = $y_intercept + $slope_mean * $x_q2;
    
    $stats{$cluster}{calculated} = {
        mean_point => [$x_mean, $calc_y_mean],
        median_point => [$x_q2, $calc_y_median],
	n => $m
    };
}

my $stats_file = 'output_EZ/group_statistics.txt';
open my $out_fh, '>', $stats_file or die "Cannot open file: $!";
foreach my $cluster (sort { $a <=> $b } keys %stats) {
    print $out_fh "Statistics for Group $cluster:\n";
    
    printf $out_fh "X: Mean=%.4f, Q1=%.4f, Q2=%.4f, Q3=%.4f\n",
        $stats{$cluster}{x}{mean},
        $stats{$cluster}{x}{q1},
        $stats{$cluster}{x}{q2},
        $stats{$cluster}{x}{q3};

    printf $out_fh "Y: Mean=%.4f, Q1=%.4f, Q2=%.4f, Q3=%.4f\n",
        $stats{$cluster}{y}{mean},
        $stats{$cluster}{y}{q1},
        $stats{$cluster}{y}{q2},
        $stats{$cluster}{y}{q3};
    
    printf $out_fh "Slope: Mean=%.4f, Q1=%.4f, Q2=%.4f, Q3=%.4f\n",
        $stats{$cluster}{slope}{mean},
        $stats{$cluster}{slope}{q1},
        $stats{$cluster}{slope}{q2},
        $stats{$cluster}{slope}{q3};
    
    print $out_fh "\n";
}
close $out_fh;

my $calc_file = 'output_EZ/calculated_points.txt';
open my $calc_fh, '>', $calc_file or die "Cannot open file: $!";
print $calc_fh "Group\tX_Mean\tY_Calculated_Mean\tX_Median\tY_Calculated_Median\tg_num\n";
foreach my $cluster (sort { $a <=> $b } keys %stats) {
    printf $calc_fh "%d\t%.4f\t%.4f\t%.4f\t%.4f\t%d\n",
        $cluster,
        $stats{$cluster}{calculated}{mean_point}[0],
        $stats{$cluster}{calculated}{mean_point}[1],
        $stats{$cluster}{calculated}{median_point}[0],
        $stats{$cluster}{calculated}{median_point}[1],
	$stats{$cluster}{calculated}{n};
}
close $calc_fh;


print "Input file: $input_file\n";
print "Output files: $stats_file, $calc_file\n";

File: preparation_2.py
#!/usr/bin/env python3
# Copyright (c) 2025. RIKEN All rights reserved.
# This is for academic and non-commercial research use only.
# The technology is currently under patent application.
# Commercial use is prohibited without a separate license agreement.
# E-mail: akihiro.ezoe@riken.jp

import sys
import os

def calculate_quartile_indices(n):
    if n == 0:
        return None, None, None

    q1_idx = int(n / 4)
    q2_idx = int(n / 2)
    q3_idx = int(n * 3 / 4)

    return q1_idx, q2_idx, q3_idx

def calculate_stats(values):
    n = len(values)
    if n == 0:
        return 0, 0, 0, 0, 0

    sorted_values = sorted(values)

    mean = sum(sorted_values) / n

    q1_idx, q2_idx, q3_idx = calculate_quartile_indices(n)

    q1 = sorted_values[q1_idx]
    q2 = sorted_values[q2_idx]
    q3 = sorted_values[q3_idx]

    return mean, q1, q2, q3, n

def main():
    input_file = 'output_EZ/clustered_data.txt'
    stats_file = 'output_EZ/group_statistics.txt'
    calc_file = 'output_EZ/calculated_points.txt'

    data = []
    try:
        with open(input_file, 'r') as f:
            header = f.readline()
            for line in f:
                line = line.strip()
                if not line:
                    continue
                parts = line.split()
                if len(parts) >= 3:
                    data.append([float(parts[0]), float(parts[1]), int(parts[2])])
    except FileNotFoundError:
        print(f"Error: Cannot open file: {input_file}")
        sys.exit(1)

    groups = {}
    for row in data:
        cluster = row[2]
        if cluster not in groups:
            groups[cluster] = []
        x_val = row[0]
        y_val = row[1]
        diff = y_val - x_val
        groups[cluster].append([x_val, y_val, diff])

    stats_results = {}

    for cluster in sorted(groups.keys()):
        group_data = groups[cluster]
        x_values = [p[0] for p in group_data]
        y_values = [p[1] for p in group_data]
        diff_values = [p[2] for p in group_data]

        m = len(group_data)

        x_mean, x_q1, x_q2, x_q3, _ = calculate_stats(x_values)
        y_mean, y_q1, y_q2, y_q3, _ = calculate_stats(y_values)

        sorted_by_x = sorted(group_data, key=lambda item: item[0], reverse=True)

        sorted_by_diff = sorted(group_data, key=lambda item: item[2], reverse=True)

        q1_idx, q2_idx, q3_idx = calculate_quartile_indices(m)

        if q2_idx is not None:
            q2_start = sorted_by_x[q2_idx][0]

            q2_diff = sorted_by_diff[q2_idx][2]

            calc_y_q2 = q2_start + q2_diff

            q1_start = sorted_by_x[q1_idx][0]
            q1_diff = sorted_by_diff[q1_idx][2]

            q3_start = sorted_by_x[q3_idx][0]
            q3_diff = sorted_by_diff[q3_idx][2]
        else:
            q2_start = x_q2
            q2_diff = y_q2 - x_q2
            calc_y_q2 = q2_start + q2_diff
            q1_start = x_q1
            q1_diff = y_q1 - x_q1
            q3_start = x_q3
            q3_diff = y_q3 - x_q3

        diff_mean, diff_q1, diff_q2, diff_q3, _ = calculate_stats(diff_values)

        stats_results[cluster] = {
            'x': {'mean': x_mean, 'q1': x_q1, 'q2': x_q2, 'q3': x_q3},
            'y': {'mean': y_mean, 'q1': y_q1, 'q2': y_q2, 'q3': y_q3},
            'diff': {
                'mean': diff_mean,
                'q1': diff_q1,
                'q2': diff_q2,
                'q3': diff_q3,
                'q1_trapezoid': q1_diff,
                'q2_trapezoid': q2_diff,
                'q3_trapezoid': q3_diff
            },
            'calculated': {
                'mean_point': [q2_start, calc_y_q2],
                'median_point': [q2_start, calc_y_q2],
                'n': m,
                'q2_start': q2_start,
                'q2_diff': q2_diff
            }
        }

    try:
        with open(stats_file, 'w') as f:
            for cluster in sorted(stats_results.keys()):
                s = stats_results[cluster]
                f.write(f"Statistics for Group {cluster}:\n")
                f.write(f"X: Mean={s['x']['mean']:.4f}, Q1={s['x']['q1']:.4f}, Q2={s['x']['q2']:.4f}, Q3={s['x']['q3']:.4f}\n")
                f.write(f"Y: Mean={s['y']['mean']:.4f}, Q1={s['y']['q1']:.4f}, Q2={s['y']['q2']:.4f}, Q3={s['y']['q3']:.4f}\n")
                f.write(f"Diff: Mean={s['diff']['mean']:.4f}, Q1={s['diff']['q1']:.4f}, Q2={s['diff']['q2']:.4f}, Q3={s['diff']['q3']:.4f}\n")
                f.write(f"Diff(trapezoid): Q1={s['diff']['q1_trapezoid']:.4f}, Q2={s['diff']['q2_trapezoid']:.4f}, Q3={s['diff']['q3_trapezoid']:.4f}\n")
                f.write(f"Calculated: Q2s={s['calculated']['q2_start']:.4f}, Q2d={s['calculated']['q2_diff']:.4f}\n")
                f.write("\n")
    except IOError as e:
        print(f"Error writing to {stats_file}: {e}")
        sys.exit(1)

    try:
        with open(calc_file, 'w') as f:
            f.write("Group\tX_Mean\tY_Calculated_Mean\tX_Median\tY_Calculated_Median\tg_num\n")
            for cluster in sorted(stats_results.keys()):
                s = stats_results[cluster]
                c = s['calculated']
                f.write(f"{cluster}\t{c['mean_point'][0]:.4f}\t{c['mean_point'][1]:.4f}\t{c['median_point'][0]:.4f}\t{c['median_point'][1]:.4f}\t{c['n']}\n")
    except IOError as e:
        print(f"Error writing to {calc_file}: {e}")
        sys.exit(1)

    print(f"Input file: {input_file}")
    print(f"Output files: {stats_file}, {calc_file}")
    print(f"\nCalculation method: trapezoid (Q2s, Q2s + Q2d)")
    print(f"  Q2s = Median of X values (sorted by X, descending)")
    print(f"  Q2d = Median of diff values (sorted by diff, descending)")

if __name__ == "__main__":
    main()
